### ISSUE TITLE
Mamba encountering problem while solving ISIS dependencies

### ISSUE THREAD

**Issue by AsterAphelion** [DOI-USGS/ISIS3#4633]
**ISIS version(s) affected**: 6.0.0

**Description**  
Mamba encounters the following problem while solving for installation dependencies when setting up a new ISIS environment in conda. 

> Encountered problems while solving:
>   - package libnghttp2-1.43.0-h812cca2_0 requires openssl >=1.1.1i,<1.1.2a, but none of the providers can be installed

**How to reproduce**  
This issue occurs when following the setup instructions on the project page.

**Comment by jlaura** [comment]
@AsterAphelion What OS/version are you installing onto? What does your .condarc look like (specifically what channels and their order)?

**Comment by astrostu** [comment]
Same problem, macOS 10.15.7.  I don't know how to access the .condarc you mention, but the channels order is correct.

I've tried various incarnations of the above.  Trying to just install "isis" gives me version 3.6.something.  isis=5.0.0, 5.0.1, 5.0.2, or 6.0 always gives me incompatibilities similar to the above.  Tried custom installing version 1.39 of the libnghttp but still got the requirement on openssl.  Tried using different versions of python, 3.6, 3.6.7 (what I have isis 4.4.0 running from), all similar incompatibility errors.

**Comment by AsterAphelion** [comment]
I am running Linux Mint 20.2. I initially attempted to install ISIS 6.0.0, but the same error occurred with all version 5 releases as well.

My .condarc file for the isis environment is the following:

channels:
  - usgs-astrogeology
  - conda-forge
  - defaults

EDIT: Also forgot to add that when mamba did not work, I was able to successfully install 6.0.0 using the conda solver.

**Comment by astrostu** [comment]
I got ISIS 6.0.0 to solve via conda overnight and installed it.  Can't run it, says it can't find libxerces-c-3.1.dylib, I'm making a new issue on that.  Conda is still trying to solve 5.0.2 ... it's been about 20 hours.

**Comment by foobarbecue** [comment]
I had this issue too. I'm on WSL2, latest miniconda, all installed fresh today. When I switched from using mamba to using conda, installation started fine (although it took a few minutes to solve). It's installing now so I'll post once I get a chance to verify that it works.

**Comment by foobarbecue** [comment]
Installed, but nothing runs. All isis commands fail with:
`lronac2isis: error while loading shared libraries: libssh2.so.1: cannot open shared object file: No such file or directory`

**Comment by foobarbecue** [comment]
I did `mamba install libssh2` and now it wants `libavresample`. Seems like recent isis packages do not fully specify their dependancies.

**Comment by foobarbecue** [comment]
Now it's looking for libarmadillo.so.9 . I guess it's this problem: https://github.com/USGS-Astrogeology/ISIS3/issues/4108

**Comment by astrostu** [comment]
Yes ... it seems like the move to recommend mamba might've been slightly premature.  I hope the team is able to resolve this, I'm still trying to solve the environment for 5.0.2 with conda ...

**Comment by foobarbecue** [comment]
Hm, well I started from a fresh conda environment, following the instructions except for the mamba part, and was able to install isis 6 in about 10 minutes and everything seems to be working.

**Comment by jlaura** [comment]
@foobarbecue Glad to hear that it is working and thanks for all the contributors for posting through as they work on the installation. Some upcoming changes to use more conda-forge packages should hopefully improve the installation experience.

One potential issue is that folks have cached version of libraries locally that are causing issues. Starting from a fresh environment [after clearing package caches](https://docs.conda.io/projects/conda/en/latest/commands/clean.html) could be one solution. When we test our builds, we test in containers with clean conda environments (as it is impossible to test every possible combination of a user's local cache).

[Here](https://astroservices.usgs.gov/jenkins/job/USGS-Astrogeology/job/ISIS3/job/dev/) you can see the CI builds from PRs. The edrget failures are expected and we should have a deprecation notification coming out soon. The [ubuntu build](https://astroservices.usgs.gov/jenkins/job/USGS-Astrogeology/job/ISIS3/job/dev/271/execution/node/51/log/) might be the best one to look at (CentOS is using GCC4 and Mac had a boost build issue on the last PR). You can see conda being used in a clean environment and resolving as expected.

I can report that a clean env also builds using mamba locally on my Mac. A good path forward would be to hear other's experiences pulling the library on a system with cleaned conda caches. That will help the packager maintainers understand if this is an issue with the build or an issue with the README that can be updated with an explicit `conda clean`.

**Comment by astrostu** [comment]
@jlaura "conda clean" ... what?  I run that and it says it needs more arguments.  To be clear, is this to be run from the base or from a new environment (and if the latter, would creating a new environment accomplish the same thing).

I'll update my note that I got 5.0.2 to finally solve, but I got my same library issue noted above, and posted the update in that issue.

**Comment by jlaura** [comment]
[conda clean docs](https://docs.conda.io/projects/conda/en/latest/commands/clean.html) describe the clean command (as previously linked).

**Comment by astrostu** [comment]
> [conda clean docs](https://docs.conda.io/projects/conda/en/latest/commands/clean.html) describe the clean command (as previously linked).

Yes, I've read that, but I don't know what arguments you are suggesting to try to make this work.  Pretend I know nothing about conda other than copy-pasting commands here and on the ISIS install page ... except it's not a pretend, it's a true.  Do you want me to run conda clean -a to remove all index cache, lock files, etc?  Just -i to remove the index cache?  -p to remove unused packages?  etc.  Is this run from the base conda environment, or from the isis one I'm trying to make?

**Comment by AndrewAnnex** [comment]
I am also seeing this issue in my CI tests for asap https://github.com/AndrewAnnex/asap_stereo/runs/3633489897, running `ldd 'which qview'` (with \` instead of ') shows that libavresample is missing, despite a using a yaml with only isis>=6.0.0 specified with the appropriate channels. 

For what it's worth, the SAT solver in Mamba is a very robust program, the fact that the conda solver makes a workable install but only after 20 hours is more of a critic of conda's solver in my book than a failure in mamba. #4634 should go a long way to make solves easier for isis.

**Comment by ryodohemmi** [comment]
I also encountered the same error when installing ISIS 5.0.1 via mamba.
My simple workaround is to create ISIS's environment stating python=3.6.6 (or >=3.6.0,<3.6.7), rather than python=3.6 (which automatically installs python 3.6.13 with openssl-1.1.1l).
`conda create -n isis501 python=3.6.6`
This is not tested for ISIS 6.0.0 or above, but may help this issue temporarily.

**Comment by steo85it** [comment]
> I also encountered the same error when installing ISIS 5.0.1 via mamba.
> My simple workaround is to create ISIS's environment stating python=3.6.6 (or >=3.6.0,<3.6.7), rather than python=3.6 (which automatically installs python 3.6.13 with openssl-1.1.1l).
> `conda create -n isis501 python=3.6.6`
> This is not tested for ISIS 6.0.0 or above, but may help this issue temporarily.

This "way around" also works with isis@6.0.0 - thanks for suggesting it!

**Comment by michaelaye** [comment]
I can report that mamba needs the resolver setting `flexible` to find everything, which you set like this:

`conda config --set channel_priority flexible`

In below log you can see how the dependency resolution fails using the other option `strict`, but works fine using `flexible`.
This was on Linux. I can well imagine that on macOS a different package is first found to not be satisfiable due to being as slightly different built process.

Interestingly, in `strict` mode, mamba first looks to satisfy isis 3.10, while in `flexible` mode it definitely installs isis 6, even when just doing `mamba install isis`  (note that no channel name needs to be given when the usgs channel is the highest in priority).

I warned about this issue one year ago in #4199 . 

Here's my current log:

```bash
❯ conda config --show-sources
==> /home/maye/miniconda3/.condarc <==
channels:
  - conda-forge

==> /home/maye/.condarc <==
channel_priority: flexible
report_errors: True

==> /home/maye/miniconda3/envs/isis6/.condarc <==
channels:
  - usgs-astrogeology
  - conda-forge


~ via 🅒 isis6 
❯ conda config --set channel_priority strict

~ via 🅒 isis6 
❯ mamba install isis

Looking for: ['isis']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache
usgs-astrogeology/linux- [====================] (00m:00s) No change
usgs-astrogeology/noarch [====================] (00m:00s) No change

Pinned packages:
  - python 3.6.*


Encountered problems while solving:
  - nothing provides jpeg 9b needed by isis-3.10.0-py36_0


~ via 🅒 isis6 
❯ conda config --set channel_priority flexible

~ via 🅒 isis6 
❯ mamba install isis

Looking for: ['isis']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache
usgs-astrogeology/noarch [====================] (00m:00s) No change
usgs-astrogeology/linux- [====================] (00m:00s) No change

Pinned packages:
  - python 3.6.*


Transaction

  Prefix: /home/maye/miniconda3/envs/isis6

  Updating specs:

   - isis
   - ca-certificates
   - certifi
   - openssl


  Package                   Version  Build                        Channel                          Size
─────────────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────────────────────────────

  + ale                       0.8.5  py36h605e78d_3               conda-forge/linux-64           Cached
  + armadillo                10.7.5  h7c03176_0                   conda-forge/linux-64           Cached
  + arpack                    3.7.0  hc6cf775_2                   conda-forge/linux-64           Cached
```

**Comment by michaelaye** [comment]
Note that this is maybe not precisely the same issue, as the OP (or a commenter) was able to install using conda and no other settings change. I thought it was worth mentioning though that this setting can break dependency resolution.

**Comment by AustinSanders** [comment]
I'm no longer able to replicate this problem after following the install instructions on the [repo](https://github.com/USGS-Astrogeology/ISIS3).  Following the installation, I do run into the libgsl issue listed in #4877 .

I'm marking this issue closed for now, as it seems to have been resolved via the dependency work that occurred earlier this year. If users still run into this issue _in a clean environment_, please reopen this issue and tag me in the comments.
